src/webgpu/shader/execution/robust_access_vertex.spec.ts

```diff
@@ -44,7 +44,430 @@ Vertex buffer contents could be randomized to prevent the case where a previous
+  }
+
+  // Bind all vertex buffers generated
+  private BindVertexBuffers(pass: GPURenderPassEncoder) {
+    let currSlot = 0;
+    for (let i = 0; i < this.vertexBuffers.length; i++) {
+      for (let j = 0; j < this.slotsPerBuffer; j++) {
+        pass.setVertexBuffer(currSlot++, this.vertexBuffers[i]);
+      }
+    }
+  }
+
+  // Create a vertex buffer from |vertexArray|
+  private GenerateVertexBuffer(vertexArray: Float32Array): GPUBuffer {
+    const [vertexBuffer, vertexMapping] = this.device.createBufferMapped({
+      size: vertexArray.byteLength,
+      usage: GPUBufferUsage.VERTEX,
+    });
+    new Float32Array(vertexMapping).set(vertexArray);
+    vertexBuffer.unmap();
+    return vertexBuffer;
+  }
+
+  // Create an index buffer from |indexArray|
+  private GenerateIndexBuffer(indexArray: Uint16Array): GPUBuffer {
+    const [indexBuffer, indexMapping] = this.device.createBufferMapped({
+      size: indexArray.byteLength,
+      usage: GPUBufferUsage.INDEX,
+    });
+    new Uint16Array(indexMapping).set(indexArray);
+    indexBuffer.unmap();
+    return indexBuffer;
+  }
+
+  // Create an indirect buffer containing draw call values
+  private GenerateIndirectBuffer(): GPUBuffer {
+    const indirectArray = new Int32Array([
+      this.vertexCount,
+      this.instanceCount,
+      this.firstVertex,
+      this.firstInstance,
+    ]);
+    const [indirectBuffer, indirectMapping] = this.device.createBufferMapped({
+      size: indirectArray.byteLength,
+      usage: GPUBufferUsage.INDIRECT,
+    });
+    new Int32Array(indirectMapping).set(indirectArray);
+    indirectBuffer.unmap();
+    return indirectBuffer;
+  }
+
+  // Create an indirect buffer containing indexed draw call values
+  private GenerateIndexedIndirectBuffer(): GPUBuffer {
+    const indirectArray = new Int32Array([
+      this.indexCount,
+      this.instanceCount,
+      this.firstVertex,
+      this.baseVertex,
+      this.firstInstance,
+    ]);
+    const [indirectBuffer, indirectMapping] = this.device.createBufferMapped({
+      size: indirectArray.byteLength,
+      usage: GPUBufferUsage.INDIRECT,
+    });
+    new Int32Array(indirectMapping).set(indirectArray);
+    indirectBuffer.unmap();
+    return indirectBuffer;
+  }
+}
+
+// Parameterize different sized types
+interface Type {
+  format: GPUVertexFormat;
+  size: number;
+  validationFunc: string;
+}
+
+const typeInfoMap: { [k: string]: Type } = {
+  float: {
+    format: 'float',
+    size: 4,
+    validationFunc: 'return valid(v);',
+  },
+  vec2: {
+    format: 'float2',
+    size: 8,
+    validationFunc: 'return valid(v.x) && valid(v.y);',
+  },
+  vec3: {
+    format: 'float3',
+    size: 12,
+    validationFunc: 'return valid(v.x) && valid(v.y) && valid(v.z);',
+  },
+  vec4: {
+    format: 'float4',
+    size: 16,
+    validationFunc: `return valid(v.x) && valid(v.y) && valid(v.z) && valid(v.w) ||
+                            v.x == 0 && v.y == 0 && v.z == 0 && (v.w == 0.0 || v.w == 1.0);`,
```

Comment:
does the (0, 0, 0, x) rule only apply to vec4s? Or can a vec3 also be (0, 0, x) ?

Reply:
AFAIK the spec says (0,0,0,x) specifically, so I assume this is only vec4.